src/feature/extraction.cc

```diff
@@ -53,6 +53,33 @@ void ScaleKeypoints(const Bitmap& bitmap, const Camera& camera,
   }
 }
 
+void MaskKeypoints(const Bitmap& mask, FeatureKeypoints* keypoints,
+                   FeatureDescriptors* descriptors) {
+  unsigned int out_index = 0;
+  BitmapColor<uint8_t> color;
+  for (unsigned int i = 0; i < keypoints->size(); ++ i) {
+    if (mask.GetPixel(static_cast<int>(keypoints->at(i).x),
+                      static_cast<int>(keypoints->at(i).y),
+                      &color) &&
+        color.r == 0) {
```

Comment:
I feel like, if mask.GetPixel() == false, we also want to exclude it here? What do you think?

Reply:
If the mask size differs from the image size, maybe a warning should be printed? The user might have mixed up the assignment of images to masks.

Reply:
That’s a good idea. The other issue that I had in mind is that SIFT can also detect features outside of the image sometimes. In any case, we should also check for valid masks as you suggested.